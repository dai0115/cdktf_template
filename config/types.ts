import { VpcConfig } from "@cdktf/provider-aws/lib/vpc";

import { S3BucketConfig } from "@cdktf/provider-aws/lib/s3-bucket";
import { S3BucketAclConfig } from "@cdktf/provider-aws/lib/s3-bucket-acl";
import { S3BucketPublicAccessBlockConfig } from "@cdktf/provider-aws/lib/s3-bucket-public-access-block";
import { Wafv2IpSetConfig } from "@cdktf/provider-aws/lib/wafv2-ip-set";
import { Wafv2RuleGroupConfig } from "@cdktf/provider-aws/lib/wafv2-rule-group";
import { Wafv2WebAclConfig } from "@cdktf/provider-aws/lib/wafv2-web-acl";
import { CloudfrontDistributionConfig } from "@cdktf/provider-aws/lib/cloudfront-distribution";
import { CloudfrontOriginAccessControlConfig } from "@cdktf/provider-aws/lib/cloudfront-origin-access-control";
import { CognitoUserPoolConfig } from "@cdktf/provider-aws/lib/cognito-user-pool";
import { EcrRepositoryConfig } from "@cdktf/provider-aws/lib/ecr-repository";
import { EcsClusterConfig } from "@cdktf/provider-aws/lib/ecs-cluster";
import { EcsServiceConfig } from "@cdktf/provider-aws/lib/ecs-service";
import { EcsTaskDefinitionConfig } from "@cdktf/provider-aws/lib/ecs-task-definition";
import {
  CodedeployDeploymentGroupAutoRollbackConfiguration,
  CodedeployDeploymentGroupBlueGreenDeploymentConfig,
  CodedeployDeploymentGroupDeploymentStyle,
} from "@cdktf/provider-aws/lib/codedeploy-deployment-group";
import { IamRoleConfig } from "@cdktf/provider-aws/lib/iam-role";

export type stageType = "dev" | "staging" | "prduction";

type prefix = `application-${stageType}`;

// Stack毎に可変なプロパティを定義
export type ConfigType = {
  region: string;
  prefix: prefix;
  stage: stageType;
  network: NetworkConfig;
  db: DbConfig;
  cognito: CognitoConfigType;
  staticWebsiteHosting: StaticWebsiteHostingConfigType;
  computing: ComputingConfigType;
};

type NetworkConfig = {
  vpcConfig: VpcConfig;
};

type DbConfig = {
  dbMasterUsername: string;
  dbFamily: string;
  clusterEngine: string;
  clusterEngineVersion: string;
  instanceCount: number;
  instanceClass: string;
};

type OptionalCognitoConfig =
  | "aliasAttributes"
  | "deviceConfiguration"
  | "emailVerificationMessage"
  | "emailVerificationSubject"
  | "smsAuthenticationMessage"
  | "smsConfiguration"
  | "smsVerificationMessage"
  | "tags"
  | "userAttributeUpdateSettings"
  | "userPoolAddOns"
  | "verificationMessageTemplate";

type CognitoConfigType = Omit<CognitoUserPoolConfig, OptionalCognitoConfig>;

type StaticWebsiteHostingConfigType = {
  s3: S3ConfigType;
  bucketAcl: S3AclType;
  ipSet: IPSetType;
  ruleGroup: RuleGroupConfigType;
  wafv2WebAcl: Wafv2WebAclType;
  cloudfrontOAC: CloudfrontOriginAccessControlConfigType;
  cloudfrontDistribution: CloudfrontDistributionWithoutType;
  s3PublicAccessBlock: S3PublicAccessBlockType;
};

// SPA（App）用S3バケットS3の型定義
type OptionalS3BucketConfig = "none";
type OptionalS3BucketAclConfig = "bucket";
type S3ConfigType = Omit<S3BucketConfig, OptionalS3BucketConfig>;
type S3AclType = Omit<S3BucketAclConfig, OptionalS3BucketAclConfig>;

// IPSetの型定義
type OptionalIpSetConfig = "provider";
type IPSetType = Omit<Wafv2IpSetConfig, OptionalIpSetConfig>;

// WafRuleGroupの型定義
type OptionalRuleGroupConfig = "name";
type RuleGroupConfigType = Omit<Wafv2RuleGroupConfig, OptionalRuleGroupConfig>;

// WafAclの型定義
type OptinalWafv2WebAclConfig = "name" | "provider";
type Wafv2WebAclType = Omit<Wafv2WebAclConfig, OptinalWafv2WebAclConfig>;

// cloudfrontOACの型定義
type OptinalCloudfrontOriginAccessControlConfig = "name";
type CloudfrontOriginAccessControlConfigType = Omit<
  CloudfrontOriginAccessControlConfig,
  OptinalCloudfrontOriginAccessControlConfig
>;

// Cloudfront Distributionの型定義
type OptionalCloudfrontDistributionType = "origin" | "defaultCacheBehavior";
type CloudfrontDistributionWithoutType = Omit<
  CloudfrontDistributionConfig,
  OptionalCloudfrontDistributionType
>;

type OptinalS3BucketPublicAccessBlockConfig = "bucket";
type S3PublicAccessBlockType = Omit<
  S3BucketPublicAccessBlockConfig,
  OptinalS3BucketPublicAccessBlockConfig
>;

type ComputingConfigType = {
  ecrConfig: Omit<EcrRepositoryConfig, "name">;
  clusterConfig: Omit<EcsClusterConfig, "name">;
  taskRoleConfig: Omit<IamRoleConfig, "name" | "assumeRolePolicy">;
  executionRoleConfig: Omit<IamRoleConfig, "name" | "assumeRolePolicy">;
  taskDefinitionConfig: Omit<
    EcsTaskDefinitionConfig,
    | "name"
    | "family"
    | "taskRoleArn"
    | "executionRoleArn"
    | "containerDefinitions"
  >;
  serviceConfig: Omit<
    EcsServiceConfig,
    | "name"
    | "cluster"
    | "taskDefinition"
    | "networkConfiguration"
    | "loadBalancer"
  >;
  autoRollbackConfig: CodedeployDeploymentGroupAutoRollbackConfiguration;
  bluegreenDeploymentConfig: CodedeployDeploymentGroupBlueGreenDeploymentConfig;
  deploymentStyleConfig: CodedeployDeploymentGroupDeploymentStyle;
  deploymentConfigName: string;
};
