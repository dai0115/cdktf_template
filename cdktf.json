{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "8f6cdb9b-f601-410e-ace2-9b4a7e619262",
  "sendCrashReports": "false",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
