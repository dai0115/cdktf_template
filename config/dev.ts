import { ConfigType } from "./types";

export const devConfig: ConfigType = {
  region: "ap-northeast-1",
  stage: "dev",
  prefix: "application-dev",
  network: {
    vpcConfig: {
      cidrBlock: "10.0.0.0/16",
      enableDnsHostnames: true,
      enableDnsSupport: true,
    },
  },
};
