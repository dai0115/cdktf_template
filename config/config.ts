import { ConfigType, stageType } from "./types";
import { devConfig } from "./dev";

// TODO stagingとproductionを作る際に@ts-ignoreを削除する
const configMap: { [key in stageType]: ConfigType } = {
  dev: devConfig,
  // @ts-ignore
  staging: {},
  // @ts-ignore
  prduction: {},
};

export const getConfig = (stage: stageType): ConfigType => {
  return configMap[stage];
};
