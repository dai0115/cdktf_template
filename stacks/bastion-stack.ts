import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";

import { SecurityGroup } from "@cdktf/provider-aws/lib/security-group";
import { EcrRepository } from "@cdktf/provider-aws/lib/ecr-repository";
import { EcsTaskDefinition } from "@cdktf/provider-aws/lib/ecs-task-definition";
import { EcsCluster } from "@cdktf/provider-aws/lib/ecs-cluster";
import { EcsService } from "@cdktf/provider-aws/lib/ecs-service";
import { DataAwsIamPolicyDocument } from "@cdktf/provider-aws/lib/data-aws-iam-policy-document";
import { IamRole } from "@cdktf/provider-aws/lib/iam-role";
import { DataAwsCallerIdentity } from "@cdktf/provider-aws/lib/data-aws-caller-identity";

import { ConfigType } from "../config/types";

type BastionProps = {
  config: ConfigType;
  bastionSubnetId: string;
  bastionSG: SecurityGroup;
};

export class BastionStack extends TerraformStack {
  constructor(scope: Construct, id: string, props: BastionProps) {
    super(scope, id);
    const { bastionSG, bastionSubnetId } = props;
    const { region, prefix } = props.config;

    new AwsProvider(this, `${prefix}-AWS`, {
      region: region,
    });

    const accountId = new DataAwsCallerIdentity(this, "account-id").accountId;

    // 踏み台のイメージを保存するためのリポジトリを作成
    const repo = new EcrRepository(this, "ecr", {
      name: `${prefix}-ecr-bastion-repo`,
      imageScanningConfiguration: { scanOnPush: true },
      encryptionConfiguration: [
        {
          encryptionType: "KMS",
        },
      ],
    });

    const taskAssumeRole = this.createAssumeRolePolicy(
      `${prefix}-task`,
      "ecs-tasks.amazonaws.com"
    );

    // タスクロールの作成
    const taskRole = new IamRole(this, `${prefix}-task-role`, {
      name: `${prefix}-task-role`,
      assumeRolePolicy: taskAssumeRole.json,
      inlinePolicy: [
        {
          name: "taskRole",
          policy: JSON.stringify({
            Version: "2012-10-17",
            Statement: [
              {
                Effect: "Allow",
                Action: [
                  "ssmmessages:CreateActivation",
                  "ssmmessages:RemoveTagsFromResource",
                  "ssmmessages:AddTagsFromResource",
                  "ssmmessages:DeleteActivation",
                ],
                Resource: "*",
              },
            ],
          }),
        },
      ],
    });
    // タスク実行ロールの作成
    const taskExecutionRole = new IamRole(
      this,
      `${prefix}-task-execution-role`,
      {
        name: `${prefix}-task-execution-role`,
        assumeRolePolicy: taskAssumeRole.json,
        inlinePolicy: [
          {
            name: "taskExecutionRole",
            policy: JSON.stringify({
              Version: "2012-10-17",
              Statement: [
                {
                  Effect: "Allow",
                  Action: [
                    "ecr:GetAuthorizationToken",
                    "ecr:BatchCheckLayerAvailability",
                    "ecr:GetDownloadUrlForLayer",
                    "ecr:BatchGetImage",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                    "secretsmanager:GetSecret",
                    "secretsmanager:GetSecretValue",
                    "secretsmanager:DescribeSecret",
                    "secretsmanager:ListSecrets",
                    "logs:*",
                  ],
                  Resource: "*",
                },
              ],
            }),
          },
        ],
      }
    );

    // ECSタスクがssmにわたすIAMロールの作成
    const ssmAssumeRole = this.createAssumeRolePolicy(
      `${prefix}-ssm`,
      "ssm.amazonaws.com"
    );
    // ssmが利用するロールの作成
    new IamRole(this, `${prefix}-ssm-role`, {
      name: `${prefix}-ssm-role`,
      assumeRolePolicy: ssmAssumeRole.json,
      managedPolicyArns: [
        "arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore",
      ],
    });

    // ECS関連のリソースを作成
    const cluster = new EcsCluster(this, `${prefix}-BastionCluster`, {
      name: `${prefix}-BastionCluster`,
      setting: [
        {
          name: "containerInsights",
          value: "enabled",
        },
      ],
    });

    const taskDefinition = new EcsTaskDefinition(
      this,
      `${prefix}-BastiontaskDefinition`,
      {
        family: `${prefix}-BastiontaskDefinition`,
        taskRoleArn: taskRole.arn,
        executionRoleArn: taskExecutionRole.arn,
        requiresCompatibilities: ["FARGATE"],
        cpu: "256",
        memory: "512",
        networkMode: "awsvpc",
        containerDefinitions: JSON.stringify([
          {
            name: "bastion",
            image: `${accountId}.dkr.ecr.ap-northeast-1.amazonaws.com/${repo.name}:latest`,
            essential: true,
            cpu: 256,
            memory: 512,
            portMappings: [
              {
                containerPort: 80,
              },
            ],
            logConfiguration: {
              logDriver: "awslogs",
              options: {
                "awslogs-region": "ap-northeast-1",
                "awslogs-stream-prefix": "ecs-bastion-task",
                "awslogs-group": "/ecs/ecs-bastion-task",
                "awslogs-create-group": "true",
              },
            },
          },
        ]),
      }
    );

    new EcsService(this, `${prefix}-BastionEcsService`, {
      name: `${prefix}-BastionEcsService`,
      cluster: cluster.name,
      enableExecuteCommand: true,
      taskDefinition: taskDefinition.arn,
      desiredCount: 1,
      launchType: "FARGATE",
      networkConfiguration: {
        securityGroups: [bastionSG.id],
        subnets: [bastionSubnetId],
      },
    });
  }

  /**
   * @param prefix - リソース命名用のprefix
   * @param identifiers - アクセスするAWSサービスのidentifiersを指定
   * @returns 作成したロールをJSON形式で返却
   */
  private createAssumeRolePolicy(
    prefix: string,
    identifiers: string
  ): DataAwsIamPolicyDocument {
    const assumeRolePolicy = new DataAwsIamPolicyDocument(
      this,
      `${prefix}-assumeRolePolicy`,
      {
        statement: [
          {
            effect: "Allow",
            actions: ["sts:AssumeRole"],
            principals: [
              {
                identifiers: [identifiers],
                type: "Service",
              },
            ],
          },
        ],
      }
    );
    return assumeRolePolicy;
  }
}
