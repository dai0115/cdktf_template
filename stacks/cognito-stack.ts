import { CognitoUserPool } from "@cdktf/provider-aws/lib/cognito-user-pool";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";
import { ConfigType } from "../config/types";

type CognitoProps = {
  config: ConfigType;
};

export class CognitoStack extends TerraformStack {
  readonly cognitoId: String;

  constructor(scope: Construct, id: string, props: CognitoProps) {
    super(scope, id);
    const { region, prefix } = props.config;
    const { cognito } = props.config;

    new AwsProvider(this, `${prefix}-AWS`, {
      region: region,
    });

    const cog = new CognitoUserPool(this, `${prefix}-Cognito`, {
      ...cognito,
    });
    this.cognitoId = cog.id;
  }
}
