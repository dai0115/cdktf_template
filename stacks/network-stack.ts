import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { TerraformStack } from "cdktf";
import { Construct } from "constructs";

import { Vpc } from "@cdktf/provider-aws/lib/vpc";
import { Subnet } from "@cdktf/provider-aws/lib/subnet";
import { SecurityGroup } from "@cdktf/provider-aws/lib/security-group";
import { SecurityGroupRule } from "@cdktf/provider-aws/lib/security-group-rule";
import { InternetGateway } from "@cdktf/provider-aws/lib/internet-gateway";
import { RouteTable } from "@cdktf/provider-aws/lib/route-table";
import { RouteTableAssociation } from "@cdktf/provider-aws/lib/route-table-association";
import { VpcEndpoint } from "@cdktf/provider-aws/lib/vpc-endpoint";

import { ConfigType } from "../config/types";

export class NetworkStack extends TerraformStack {
  readonly ecsSubnetIds: Array<string>;
  readonly dbSubnetIds: Array<string>;
  readonly albSubnetIds: Array<string>;
  readonly endpointSubnetId: string;
  readonly albSG: SecurityGroup;
  readonly ecsSG: SecurityGroup;
  readonly dbSG: SecurityGroup;

  constructor(scope: Construct, id: string, props: ConfigType) {
    super(scope, id);

    const { region, prefix } = props;
    const { vpcConfig } = props.network;

    new AwsProvider(this, `${prefix}-AWS`, {
      region: region,
    });

    // VPCリソースの作成
    const vpc = new Vpc(this, `${prefix}-Vpc`, {
      ...vpcConfig,
      tags: {
        Name: `${prefix}-Vpc`,
      },
    });

    // サブネットの作成
    const ecsSubnet1 = this.createSubnet(
      `${prefix}-Ecs1`,
      vpc.id,
      "10.0.0.0/24",
      "ap-northeast-1a"
    );
    const ecsSubnet2 = this.createSubnet(
      `${prefix}-Ecs2`,
      vpc.id,
      "10.0.1.0/24",
      "ap-northeast-1c"
    );
    this.ecsSubnetIds = [ecsSubnet1, ecsSubnet2];
    const dbSubnet1 = this.createSubnet(
      `${prefix}-Db1`,
      vpc.id,
      "10.0.2.0/24",
      "ap-northeast-1a"
    );
    const dbSubnet2 = this.createSubnet(
      `${prefix}-Db2`,
      vpc.id,
      "10.0.3.0/24",
      "ap-northeast-1c"
    );
    this.dbSubnetIds = [dbSubnet1, dbSubnet2];
    const albSubnet1 = this.createSubnet(
      `${prefix}-Alb1`,
      vpc.id,
      "10.0.4.0/24",
      "ap-northeast-1a"
    );
    const albSubnet2 = this.createSubnet(
      `${prefix}-Alb2`,
      vpc.id,
      "10.0.5.0/24",
      "ap-northeast-1c"
    );
    this.albSubnetIds = [albSubnet1, albSubnet2];
    this.endpointSubnetId = this.createSubnet(
      `${prefix}-Endpoint`,
      vpc.id,
      "10.0.6.0/24",
      "ap-northeast-1a"
    );
    const bastionSubnetId = this.createSubnet(
      `${prefix}-Bastion`,
      vpc.id,
      "10.0.7.0/24",
      "ap-northeast-1a"
    );

    // ルートテーブルの作成
    const ecsRouteTable = this.createRouteTable(`${prefix}-EcsRoute`, vpc);
    const dbRouteTable = this.createRouteTable(`${prefix}-DbRoute`, vpc);
    const albRouteTable = this.createRouteTable(`${prefix}-AlbRoute`, vpc);
    const bastionRouteTable = this.createRouteTable(
      `${prefix}-BastionRoute`,
      vpc
    );
    const endpointRouteTable = this.createRouteTable(
      `${prefix}-EndpointRoute`,
      vpc
    );

    // サブネットとルートテーブルの関連付け
    this.attachRouteTabletoSubnet("ecs1", ecsSubnet1, ecsRouteTable.id);
    this.attachRouteTabletoSubnet("ecs2", ecsSubnet2, ecsRouteTable.id);
    this.attachRouteTabletoSubnet("db1", dbSubnet1, dbRouteTable.id);
    this.attachRouteTabletoSubnet("db2", dbSubnet2, dbRouteTable.id);
    this.attachRouteTabletoSubnet("alb1", albSubnet1, albRouteTable.id);
    this.attachRouteTabletoSubnet("alb2", albSubnet2, albRouteTable.id);
    this.attachRouteTabletoSubnet(
      "bastion",
      bastionSubnetId,
      bastionRouteTable.id
    );
    this.attachRouteTabletoSubnet(
      "endpoint",
      this.endpointSubnetId,
      endpointRouteTable.id
    );

    // パブリックサブネットへのリソースの関連付け
    const iGW = this.createIGW(prefix, vpc.id);
    new RouteTableAssociation(this, "ecs-IGW", {
      gatewayId: iGW.id,
      routeTableId: albRouteTable.id,
    });

    // セキュリティグループの作成
    this.albSG = this.createSecurityGroup("alb", vpc.id);
    new SecurityGroupRule(this, "albSG-ingress-rule", {
      cidrBlocks: ["0.0.0.0/0"],
      type: "ingress",
      fromPort: 80, //TODO replace here to 443
      toPort: 80, //TODO replace here to 443
      protocol: "tcp",
      securityGroupId: this.albSG.id,
      description: "IGW to Load balancer",
    });
    this.attachAllTrafficRules("alb", this.albSG.id);

    this.ecsSG = this.createSecurityGroup("ecs", vpc.id);
    new SecurityGroupRule(this, "ecsSG-ingress-rule", {
      sourceSecurityGroupId: this.albSG.id,
      type: "ingress",
      fromPort: 80,
      toPort: 80,
      protocol: "tcp",
      securityGroupId: this.ecsSG.id,
      description: "Load balancer to target",
    });
    this.attachAllTrafficRules("ecs", this.ecsSG.id);

    this.dbSG = this.createSecurityGroup("db", vpc.id);
    new SecurityGroupRule(this, "dbSG-ingress-rule2", {
      sourceSecurityGroupId: this.ecsSG.id,
      type: "ingress",
      fromPort: 3306,
      toPort: 3306,
      protocol: "tcp",
      securityGroupId: this.dbSG.id,
    });
    this.attachAllTrafficRules("db", this.dbSG.id);

    const endpointSG = this.createSecurityGroup("endpoint", vpc.id);
    new SecurityGroupRule(this, "endpointSG-ingress-rule", {
      cidrBlocks: ["10.0.0.0/16"],
      type: "ingress",
      fromPort: 443,
      toPort: 443,
      protocol: "tcp",
      securityGroupId: endpointSG.id,
    });
    this.attachAllTrafficRules("endpoint", endpointSG.id);

    const bastionSG = this.createSecurityGroup("bastion", vpc.id);
    this.attachAllTrafficRules("bastion", bastionSG.id);

    // VPCエンドポイントの作成
    this.createGatewayEndpoint("ecs", vpc.id, ecsRouteTable.id, "s3");
    this.createGatewayEndpoint("bastion", vpc.id, bastionRouteTable.id, "s3");

    const serviceNames: string[] = [
      "ssm",
      "ssmmessages",
      "ec2messages",
      "ecr.api",
      "ecr.dkr",
      "logs",
      "secretsmanager",
      "ecs-agent",
      "ecs-telemetry",
      "ecs",
    ];
    for (const service of serviceNames) {
      this.createInterfaceEndpoint(
        vpc.id,
        endpointSG.id,
        this.endpointSubnetId,
        service
      );
    }
  }

  /**
   * @param name - 作成するサブネットの名前
   * @param vpcId - サブネットを作成するvpcのId
   * @param cideBlock - 作成するサブネットのcideBlock
   * @param availabilityZone - サブネットを作成するavailabilityZone
   * @returns 作成されたSubnetを返却
   */
  private createSubnet(
    name: string,
    vpcId: string,
    cideBlock: string,
    availabilityZone: string
  ): string {
    const subnet = new Subnet(this, name, {
      vpcId: vpcId,
      cidrBlock: cideBlock,
      availabilityZone: availabilityZone,
      tags: {
        Name: name,
      },
    });
    return subnet.id;
  }

  /**
   * @param name - 作成するルートテーブルの名前
   * @param vpcId - ルートテーブルを作成するvpcのId
   * @returns 作成されたRouteTableを返却
   */
  private createRouteTable(name: string, vpc: Vpc): RouteTable {
    return new RouteTable(this, name, {
      vpcId: vpc.id,
      route: [],
      tags: {
        Name: name,
      },
    });
  }

  /**
   * @param name - 命名時に利用するprefix
   * @param subnetId - ルートテーブルと関連付けを行うサブネットのId
   * @param subnetId - サブネットと関連付けを行うルートテーブルのId
   * 受け取ったサブネットとルートテーブルを紐付ける
   */
  private attachRouteTabletoSubnet(
    name: string,
    subnetId: string,
    routeTableId: string
  ) {
    new RouteTableAssociation(this, `${name}-Association`, {
      subnetId: subnetId,
      routeTableId: routeTableId,
    });
  }

  /**
   * @param name - リソースの命名に利用するprefix
   * @param vpcId - IGWを作成するvpcのId
   * @returns 作成されたInternet Gatewayを返却
   */
  private createIGW(name: string, vpcId: string): InternetGateway {
    return new InternetGateway(this, `${name}-IGW`, {
      vpcId: vpcId,
      tags: {
        Name: `${name}-IGW`,
      },
    });
  }

  /**
   * @param name - セキュリティグループの名前
   * @param vpcId - セキュリティグループを作成するvpcのId
   */
  private createSecurityGroup(name: string, vpcId: string): SecurityGroup {
    return new SecurityGroup(this, `${name}SG`, {
      name: `${name}SG`,
      vpcId: vpcId,
      tags: {
        Name: `${name}SG`,
      },
    });
  }
  /**
   * @param name - セキュリティグループルールの名前
   * @param securityGroupId - ルールを紐付けるセキュリティグループのId
   *
   * 全ての通信を許可する場合のセキュリティグループルールを作成
   */
  private attachAllTrafficRules(name: string, securityGroupId: string) {
    new SecurityGroupRule(this, `${name}SG-egress-rule`, {
      cidrBlocks: ["0.0.0.0/0"],
      type: "egress",
      fromPort: 0,
      toPort: 0,
      protocol: "-1",
      securityGroupId: securityGroupId,
    });
  }
  /**
   * @param vpcId - Vpcエンドポイントを作成するVpcのId
   * @param routeTableIds - VpcエンドポイントGatewayタイプに紐付けるルートテーブルのId
   * @param serviceName - 作成するサービスの名前を指定。"s3"や"ssm"などの形式で"com.amazon〜"は不要
   * @returns 作成したVPCエンドポイントを返却
   *
   * ゲートウェイ型のエンドポイントを作成
   */
  private createGatewayEndpoint(
    name: string,
    vpcId: string,
    routeTableId: string,
    serviceName: string
  ): VpcEndpoint {
    return new VpcEndpoint(this, `${name}-vpc-endpoint-${serviceName}`, {
      vpcId: vpcId,
      serviceName: `com.amazonaws.ap-northeast-1.${serviceName}`,
      vpcEndpointType: "Gateway",
      routeTableIds: [routeTableId],
      tags: {
        Name: `vpc-endpoint-${serviceName}-Gateway`,
      },
    });
  }
  /**
   * @param vpcId - vpcエンドポイントを作成するVpcのId
   * @param securityGroupIds - Vpcエンドポイントに作成するセキュリティグループのId
   * @param subnetIds - Vpcエンドポイントを作成するサブネットのId
   * @param serviceName - 作成するサービスの名前を指定。"s3"や"ssm"などの形式で"com.amazon〜"は不要
   * @returns 作成したVPCエンドポイントを返却
   *
   * インターフェース型のエンドポイントを作成
   */
  private createInterfaceEndpoint(
    vpcId: string,
    securityGroupId: string,
    subnetId: string,
    serviceName: string
  ): VpcEndpoint {
    return new VpcEndpoint(this, `vpc-endpoint-${serviceName}`, {
      vpcId: vpcId,
      serviceName: `com.amazonaws.ap-northeast-1.${serviceName}`,
      vpcEndpointType: "Interface",
      subnetIds: [subnetId],
      securityGroupIds: [securityGroupId],
      privateDnsEnabled: true,
      tags: {
        Name: `vpc-endpoint-${serviceName}`,
      },
    });
  }
}
