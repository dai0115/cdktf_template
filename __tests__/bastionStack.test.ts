import { Testing } from "cdktf";
import { NetworkStack } from "../stacks/network-stack";
import { BastionStack } from "../stacks/bastion-stack";
import { devConfig } from "../config/dev";

describe("bastionStack", () => {
  const app = Testing.app();
  const vpc = new NetworkStack(app, "networkStack", devConfig);
  const bastion = new BastionStack(app, "bastionStack", {
    config: devConfig,
    bastionSG: vpc.bastionSG,
    bastionSubnetId: vpc.bastionSubnetId,
  });

  it("snapshot test for bastionStack", () => {
    expect(Testing.synth(bastion)).toMatchSnapshot();
  });
});
