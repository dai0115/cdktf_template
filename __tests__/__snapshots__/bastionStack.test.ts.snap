// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bastionStack snapshot test for bastionStack 1`] = `
"{
  "data": {
    "aws_iam_policy_document": {
      "application-dev-BastionAssumeRole": {
        "statement": [
          {
            "actions": [
              "sts:AssumeRole"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "ecs-tasks.amazonaws.com"
                ],
                "type": "Service"
              }
            ]
          }
        ]
      },
      "application-dev-BastionTaskPolicy": {
        "statement": [
          {
            "actions": [
              "ssmmessages:CreateControlChannel",
              "ssmmessages:CreateDataChannel",
              "ssmmessages:OpenControlChannel",
              "ssmmessages:OpenDataChannel"
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ]
          },
          {
            "actions": [
              "ecr:GetAuthorizationToken",
              "ecr:BatchCheckLayerAvailability",
              "ecr:GetDownloadUrlForLayer",
              "ecr:BatchGetImage",
              "logs:CreateLogStream",
              "logs:PutLogEvents"
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ]
          }
        ]
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "ap-northeast-1"
      }
    ]
  },
  "resource": {
    "aws_ecs_cluster": {
      "application-dev-BastionCluster": {
        "name": "application-dev-BastionCluster",
        "setting": [
          {
            "name": "containerInsights",
            "value": "enabled"
          }
        ]
      }
    },
    "aws_ecs_service": {
      "application-dev-BastionEcsService": {
        "cluster": "\${aws_ecs_cluster.application-dev-BastionCluster.name}",
        "desired_count": 1,
        "enable_execute_command": true,
        "launch_type": "FARGATE",
        "name": "application-dev-BastionEcsService",
        "network_configuration": {
          "security_groups": [
            "\${aws_security_group.bastionSG.id}"
          ],
          "subnets": [
            "\${aws_subnet.application-dev-Bastion.id}"
          ]
        },
        "task_definition": "\${aws_ecs_task_definition.application-dev-BastiontaskDefinition.arn}"
      }
    },
    "aws_ecs_task_definition": {
      "application-dev-BastiontaskDefinition": {
        "container_definitions": "[{\\"name\\":\\"bastion\\",\\"image\\":\\"amazonlinux:latest\\",\\"essential\\":true,\\"cpu\\":256,\\"memory\\":512,\\"portMappings\\":[{\\"containerPort\\":80}],\\"logConfiguration\\":{\\"logDriver\\":\\"awslogs\\",\\"options\\":{\\"awslogs-region\\":\\"ap-northeast-1\\",\\"awslogs-stream-prefix\\":\\"ecs-bastion-task\\",\\"awslogs-group\\":\\"/ecs/ecs-bastion-task\\",\\"awslogs-create-group\\":\\"true\\"}}}]",
        "cpu": "256",
        "execution_role_arn": "\${aws_iam_role.application-dev-BastionTaskExecutionRole.arn}",
        "family": "application-dev-BastiontaskDefinition",
        "memory": "512",
        "network_mode": "awsvpc",
        "requires_compatibilities": [
          "FARGATE"
        ],
        "task_role_arn": "\${aws_iam_role.application-dev-BastionTaskExecutionRole.arn}"
      }
    },
    "aws_iam_policy": {
      "application-dev-BastionTaskExecutionPolicy": {
        "name": "application-dev-BastionTaskExecutionPolicy",
        "policy": "\${data.aws_iam_policy_document.application-dev-BastionTaskPolicy.json}"
      }
    },
    "aws_iam_role": {
      "application-dev-BastionTaskExecutionRole": {
        "assume_role_policy": "\${data.aws_iam_policy_document.application-dev-BastionAssumeRole.json}",
        "name": "application-dev-BastionTaskExecutionRole"
      }
    },
    "aws_iam_role_policy_attachment": {
      "application-dev-RolePolicyAttach": {
        "policy_arn": "\${aws_iam_policy.application-dev-BastionTaskExecutionPolicy.arn}",
        "role": "\${aws_iam_role.application-dev-BastionTaskExecutionRole.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "4.60.0"
      }
    }
  }
}"
`;
